[wasm] Use source built emsdk packages for the runtime build

steveisok opened the pull request:
This change moves away from using upstream emsdk to our source built packages as part of the wasm build.

steveisok committed:
Use the emsdk transport packages to build wasm instead of a cloned emsdk

Works for the most part out of the box, but fails on the part where we npm install after
the libraries build. The reason for the failure is our node transport package is incomplete. If
you replace it with a legit node, the build works fully.

steveisok committed:
Pull in node transport package as opposed to the one packaged in emsdk. Streamline package deps

steveisok committed:
Make paths friendly for windows

steveisok committed:
Work in windows transport packages, copy python to its own directory, and adjust a bunch of paths

steveisok committed:
Bump to latest version of node that contains icu

steveisok committed:
Add windows deps and add DOTNET_EMSCRIPTEN_NODE_PATH because windows can't see npm without it

steveisok committed:
Bump emscripten packages to the latest

steveisok commented:
@akoeplinger I hand edited the eng/Version.xml/props. Do we have to do anything w/ darc to help resolve what's failing?

akoeplinger committed:
Fix typos in Version.Details.xml

akoeplinger commented:
It was just a typo :)

I'm wondering though whether we need to add one of these source-build .Intermediate dependencies too?

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -153,29 +187,80 @@
+    
+    <ItemGroup>
+      <EmsdkFiles   Condition="'%(PackageReference.Identity)' != 'runtime.$(_portableHostOS)-$(BuildArchitecture).Microsoft.NETCore.Runtime.Wasm.Node.Transport' and '%(PackageReference.Identity)' != 'Microsoft.NET.Runtime.Emscripten.3.1.34.Python.win-$(BuildArchitecture)'"
+                    Include="$(NuGetPackageRoot)\$([System.String]::Copy(%(PackageReference.Identity)).ToLowerInvariant())\%(PackageReference.Version)\tools\**" />
```
I think instead of NuGetPackageRoot you should be able to use GeneratePackagePath

radekdoulik replied:
I tried that and it gets complicated as the package names are based on other properties, so the generated property name is for example `PkgMicrosoft_NET_Runtime_Emscripten_3_1_34_Cache_$(_portableHostOS)-$(BuildArchitecture)`. It would work much better if the package path was added to the `PackageReference` items instead.

maraf commented on src/mono/browser/runtime/package-lock.json:
Omit these changes. You have too new tooling

cc @pavelsavara

steveisok committed:
Switch out to the plain mariner container to ensure no EMSDK already exists

steveisok committed:
Revert back to the browser-wasm docker image and bump to the latest

lewing commented:
I want to start using the packages asap but I don't think we want to use individual props for each item here when the package name needs to change whenever we do an emsdk bump.

steveisok commented:
> I want to start using the packages asap but I don't think we want to use individual props for each item here when the package name needs to change whenever we do an emsdk bump.

Change the package name out of emsdk then?

steveisok committed:
Container type-o

steveisok committed:
Fix condition where the node path isn't set properly

steveisok committed:
Provision even when building offsets

steveisok committed:
Fix goofy paths

steveisok committed:
Update new template to use the latest browser-wasm container

kg commented on eng/Versions.props:
```diff
@@ -255,5 +255,26 @@
     <!-- installer version, for testing workloads -->
     <MicrosoftDotnetSdkInternalVersion>9.0.100-preview.4.24175.5</MicrosoftDotnetSdkInternalVersion>
     <SdkVersionForWorkloadTesting>$(MicrosoftDotnetSdkInternalVersion)</SdkVersionForWorkloadTesting>
+    <MicrosoftNETRuntimeEmscripten3134Nodeosxarm64Version>9.0.0-preview.4.24175.6</MicrosoftNETRuntimeEmscripten3134Nodeosxarm64Version>
```
Could we use a common variable for all of these duplicates? It would make it stand out more visually when one of them has to be a different version for some reason

kg commented on src/mono/browser/Makefile:
```diff
@@ -167,4 +167,4 @@ build-app-host:
 
 patch-deterministic:
-	cd emsdk/upstream/emscripten/ && patch -p1 < ../../../runtime/deterministic.diff
+	cd emsdk/emscripten/ && patch -p1 < ../../../runtime/deterministic.diff
```
Please add a comment explaining why this is here, while you're touching it. The diff isn't self-explanatory

steveisok replied:
yeah, the gist is that if you clone emsdk and install, it'll pull a bunch of things and put them in the 'upstream' directory. Our source built packages don't have this directory structure.

kg replied:
Sorry, I mean that deterministic.diff isn't self-explanatory

radekdoulik replied:
it predates this PR. it was added by https://github.com/dotnet/runtime/pull/75532, not sure what was the context there. @lewing do you remember?

lewing replied:
This was something Zoltan used to make the each run deterministic so that he could do direct comparisons, it is only for development/testing

steveisok committed:
Don't put python.exe on the path for windows, but the folder instead

kg commented on src/mono/browser/runtime/package-lock.json:
```diff
@@ -1,130 +1,203 @@
   "name": "@microsoft/dotnet-runtime",
   "version": "1.0.0",
-  "lockfileVersion": 1,
+  "lockfileVersion": 3,
```
Make sure you actually intended to update the lockfile, lockfile changes get generated at random in my experience

pavelsavara replied:
if we want to stay on Node14 this must stay `"lockfileVersion": 1`
I would appreciate if that could be separate PR anyway.

radekdoulik replied:
this part was reverted

kg commented on src/mono/mono.proj:
```diff
@@ -87,6 +87,40 @@
   </ItemGroup>
 
+  <PropertyGroup Condition="'$(TargetsBrowser)' == 'true'">
+    <NodePkgVersion Condition="'$(HostOS)' == 'linux' and '$(BuildArchitecture)' == 'x64'">$(runtimelinuxx64MicrosoftNETCoreRuntimeWasmNodeTransportVersion)</NodePkgVersion>
```
https://learn.microsoft.com/en-us/visualstudio/msbuild/choose-element-msbuild?view=vs-2022 might make this easier to express and less typo prone

radekdoulik replied:
updated. it is a bit verbose though

kg commented on src/mono/mono.proj:
```diff
@@ -145,37 +179,91 @@
       <EmsdkExt Condition="'$(HostOS)' != 'windows'"></EmsdkExt>
       <EmsdkExt Condition="'$(HostOS)' == 'windows'">.ps1</EmsdkExt>
-      <EMSDK_PATH>$(ProvisionEmscriptenDir)</EMSDK_PATH>
+      <EMSDK_PATH>$(ProvisionEmscriptenDir.Replace('\','/'))</EMSDK_PATH>
```
Should the replace be host OS conditional?

steveisok commented:
@akoeplinger the browser offsets build is consistently failing, but actually works local. There shouldn't be any difference, but apparently there is. Can you look and see if you spot anything?

steveisok committed:
So that was a bad idea. May have to have emsdk_env.cmd have two entries

akoeplinger commented:
Looks like it can't find emcmake?

>  bash: line 1: emcmake: command not found

steveisok commented:
> Looks like it can't find emcmake?
> 
> > bash: line 1: emcmake: command not found

Yeah, local windows it's fine. I'm confused why CI is having trouble.

akoeplinger commented:
I can reproduce the failure using the ` mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-webassembly` docker container locally and `./build.sh -subset mono.aotcross -c release -arch wasm -os linux /p:MonoGenerateOffsetsOSGroups=browser`

steveisok commented:
> I can reproduce the failure using the ` mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-webassembly` docker container locally and `./build.sh -subset mono.aotcross -c release -arch wasm -os linux /p:MonoGenerateOffsetsOSGroups=browser`

Weird - I have an ubuntu instance locally and it's fine. Hard to imagine why it would be different on mariner?

akoeplinger commented:
Can you please try removing eng/pipelines/mono/update-machine-certs.ps1 and the step in global-build-job.yml that calls it?
I think it should no longer be needed with this PR.

radekdoulik committed:
Fix offsets generation for browser

radekdoulik committed:
Fix emsdk's python path

It is not versioned anymore as we use the content of our nugets

radekdoulik committed:
Fix EMSDK_PATH on non-windows platforms

radekdoulik committed:
Do not add link flags to compile flags

To avoid:

    C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(825,5): error : emcc: warning: linker setting ignored during compilation: 'EXPORT_ES6' [-Wunused-command-line-argument] [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
    C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(825,5): error : emcc: warning: linker setting ignored during compilation: 'EXPORT_EXCEPTION_HANDLING_HELPERS' [-Wunused-command-line-argument] [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]

radekdoulik committed:
Set the emsdk paths relative to the script location

radekdoulik committed:
Escape few characters and fix the added script

radekdoulik committed:
Fix .emscripten script

radekdoulik committed:
Use EM_CONFIG intead of __file__

radekdoulik committed:
Fix emsdk_env.cmd script

radekdoulik committed:
Revert "Do not add link flags to compile flags"

This reverts commit eb19ade3ef901669a7e6bcc944703837f26f5173.

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into test-build-wasm

radekdoulik committed:
Revert changes in package-lock.json

radekdoulik committed:
Feedback + cleaning

radekdoulik committed:
Merge branch 'main' into test-build-wasm

radekdoulik committed:
More cleaning

radekdoulik committed:
Try not to use the replace in the EMSDK_PATH

radekdoulik committed:
Merge branch 'main' into test-build-wasm

maraf commented on eng/Versions.props:
```diff
@@ -256,26 +256,28 @@
-    <runtimewinarm64MicrosoftNETCoreRuntimeWasmNodeTransportVersion>9.0.0-alpha.1.24175.1</runtimewinarm64MicrosoftNETCoreRuntimeWasmNodeTransportVersion>
-    <runtimewinx64MicrosoftNETCoreRuntimeWasmNodeTransportVersion>9.0.0-alpha.1.24175.1</runtimewinx64MicrosoftNETCoreRuntimeWasmNodeTransportVersion>
+    <MicrosoftNETRuntimeEmscripten3134Version>9.0.0-preview.4.24175.6</MicrosoftNETRuntimeEmscripten3134Version>
+    <MicrosoftNETCoreRuntimeWasmNodeTransportVersion>9.0.0-alpha.1.24175.1</MicrosoftNETCoreRuntimeWasmNodeTransportVersion>
```
Just curious, what is the scenario when these versions won't match? Are we "sometimes" producing only some packages?

radekdoulik replied:
@steveisok ^

steveisok replied:
Good question. The node transport pack comes out of dotnet/node and so it's not guaranteed to be the same. I also don't think we need to build it all the time.

lewing replied:
There will in practice only ever be one version emsdk and one for the node transfer package so I'd suggest setting those here doing something like
```xml
            <NodePkgVersion>$(runtimelinuxarm64MicrosoftNETCoreRuntimeWasmNodeTransportVersion)</NodePkgVersion>
            <EmsdkSdkVersion>$(MicrosoftNETRuntimeEmscripten3134Sdklinuxarm64Version)</EmsdkSdkVersion>
```
here and removing the platform based logic in mono.proj

lewing replied:
It is not required in the same way

lewing commented on src/mono/mono.proj:
```diff
@@ -87,6 +89,77 @@
+    </When>
+  </Choose>
+
+  <ItemGroup Condition="'$(NeedEmscriptenPackages)' == 'true'">
+    <PackageReference Include="Microsoft.NET.Runtime.Emscripten.3.1.34.Cache.$(_portableHostOS)-$(BuildArchitecture)"
+                      PrivateAssets="all"
+                      Version="$(EmsdkCacheVersion)"
+                      PackageArch="$(BuildArchitecture)" />
+    <PackageReference Include="runtime.$(_portableHostOS)-$(BuildArchitecture).Microsoft.NETCore.Runtime.Wasm.Node.Transport"
+                      Version="$(NodePkgVersion)"
+                      PackageArch="$(BuildArchitecture)" />
+    <PackageReference Include="Microsoft.NET.Runtime.Emscripten.3.1.34.Sdk.$(_portableHostOS)-$(BuildArchitecture)"
+                      PrivateAssets="all"
+                      Version="$(EmsdkSdkVersion)"
+                      PackageArch="$(BuildArchitecture)" />
+  </ItemGroup>
```
```suggestion
  <ItemGroup Condition="'$(NeedEmscriptenPackages)' == 'true'">
    <PackageReference Include="Microsoft.NET.Runtime.Emscripten.$(EmsdkVersion).Cache.$(_portableHostOS)-$(BuildArchitecture)"
                      PrivateAssets="all"
                      Version="$(EmsdkPackageVersion)"
                      PackageArch="$(BuildArchitecture)" />
    <PackageReference Include="runtime.$(_portableHostOS)-$(BuildArchitecture).Microsoft.NETCore.Runtime.Wasm.Node.Transport"
                      Version="$(NodePkgVersion)"
                      PackageArch="$(BuildArchitecture)" />
    <PackageReference Include="Microsoft.NET.Runtime.Emscripten.$(EmsdkVersion).Sdk.$(_portableHostOS)-$(BuildArchitecture)"
                      PrivateAssets="all"
                      Version="$(EmsdkPackageVersion)"
                      PackageArch="$(BuildArchitecture)" />
  </ItemGroup>
```

radekdoulik replied:
There's not `EmsdkVersion` and `EmsdkPackageVersion` properties, so that would not work. 

We can try to improve the hardcoded 3.1.34 version in the package names in a followup PR.

lewing replied:
> There's not `EmsdkVersion` and `EmsdkPackageVersion` properties, so that would not work.
> 
> We can try to improve the hardcoded 3.1.34 version in the package names in a followup PR.

I mean in combination with https://github.com/dotnet/runtime/pull/100266#discussion_r1580090404

radekdoulik committed:
Better property names

radekdoulik committed:
Use package and emsdk version properties

Co-authored-by: Larry Ewing <lewing@microsoft.com>

radekdoulik committed:
Use the updated properties

radekdoulik committed:
Fix emsdk version

radekdoulik committed:
Merge branch 'main' into test-build-wasm

radekdoulik committed:
Merge branch 'main' into test-build-wasm

lewing committed:
Merge branch 'main' into test-build-wasm

radekdoulik committed:
Merge branch 'main' into test-build-wasm

akoeplinger commented on eng/pipelines/common/templates/pipeline-with-resources.yml:
```diff
@@ -101,7 +101,7 @@ extends:
 
       browser_wasm:
-        image: mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-webassembly-20230913040940-1edc1c6
+        image: mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-webassembly
```
is there anything else left in the -webassembly image that we need or can we switch to the plain mariner image?

maraf replied:
Wasmtime and WASI SDK (which we ignore at the moment)?

akoeplinger replied:
isn't that for the wasi leg? I think this only deals with browser

radekdoulik replied:
I have opened https://github.com/dotnet/runtime/pull/102069 to try that. I am not sure whether we still run some tests with v8.

akoeplinger commented on eng/Versions.props:
```diff
@@ -256,5 +256,29 @@
     <!-- installer version, for testing workloads -->
     <MicrosoftDotnetSdkInternalVersion>9.0.100-preview.4.24215.2</MicrosoftDotnetSdkInternalVersion>
     <SdkVersionForWorkloadTesting>$(MicrosoftDotnetSdkInternalVersion)</SdkVersionForWorkloadTesting>
+    <EmsdkPackageVersion>9.0.0-preview.4.24175.6</EmsdkPackageVersion>
+    <NodePackageVersion>9.0.0-alpha.1.24175.1</NodePackageVersion>
+    <EmsdkVersion>3.1.34</EmsdkVersion>
+    <MicrosoftNETRuntimeEmscripten3134Nodeosxarm64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Nodeosxarm64Version>
+    <MicrosoftNETRuntimeEmscripten3134Sdkosxarm64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Sdkosxarm64Version>
+    <MicrosoftNETRuntimeEmscripten3134Sdkosxx64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Sdkosxx64Version>
+    <MicrosoftNETRuntimeEmscripten3134Sdkwinx64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Sdkwinx64Version>
+    <MicrosoftNETRuntimeEmscripten3134Sdkwinarm64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Sdkwinarm64Version>
+    <MicrosoftNETRuntimeEmscripten3134Sdklinuxx64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Sdklinuxx64Version>
+    <MicrosoftNETRuntimeEmscripten3134Sdklinuxarm64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Sdklinuxarm64Version>
+    <MicrosoftNETRuntimeEmscripten3134Cachelinuxx64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Cachelinuxx64Version>
+    <MicrosoftNETRuntimeEmscripten3134Cachelinuxarm64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Cachelinuxarm64Version>
+    <MicrosoftNETRuntimeEmscripten3134Cacheosxx64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Cacheosxx64Version>
+    <MicrosoftNETRuntimeEmscripten3134Cacheosxarm64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Cacheosxarm64Version>
+    <MicrosoftNETRuntimeEmscripten3134Cachewinx64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Cachewinx64Version>
+    <MicrosoftNETRuntimeEmscripten3134Cachewinarm64Version>$(EmsdkPackageVersion)</MicrosoftNETRuntimeEmscripten3134Cachewinarm64Version>
+    <runtimelinuxarm64NodePackageVersion>$(NodePackageVersion)</runtimelinuxarm64NodePackageVersion>
+    <runtimelinuxmuslarm64NodePackageVersion>$(NodePackageVersion)</runtimelinuxmuslarm64NodePackageVersion>
+    <runtimelinuxx64NodePackageVersion>$(NodePackageVersion)</runtimelinuxx64NodePackageVersion>
+    <runtimelinuxmuslx64NodePackageVersion>$(NodePackageVersion)</runtimelinuxmuslx64NodePackageVersion>
+    <runtimeosxarm64NodePackageVersion>$(NodePackageVersion)</runtimeosxarm64NodePackageVersion>
+    <runtimeosxx64NodePackageVersion>$(NodePackageVersion)</runtimeosxx64NodePackageVersion>
+    <runtimewinarm64NodePackageVersion>$(NodePackageVersion)</runtimewinarm64NodePackageVersion>
+    <runtimewinx64NodePackageVersion>$(NodePackageVersion)</runtimewinx64NodePackageVersion>
```
the node properties don't match the naming convention so I'm not sure darc will touch them.

we probably only need one property e.g. winx64 for darc to update the version?

please move this to the existing emscripten properties on L234 and check wheter the commend is still accurate

lewing replied:
```suggestion
    <runtimewinx64MicrosoftNETCoreRuntimeWasmNodeTransportPackageVersion>9.0.0-alpha.1.24175.1</runtimewinx64MicrosoftNETCoreRuntimeWasmNodeTransportPackageVersion>
    <EmsdkPackageVersion>$(MicrosoftNETRuntimeEmscriptenVersion)</EmsdkPackageVersion>
    <NodePackageVersion>$(runtimewinx64MicrosoftNETCoreRuntimeWasmNodeTransportPackageVersion)</NodePackageVersion>
    <EmsdkVersion>3.1.34</EmsdkVersion>
```

You want the darc to update the actual versions so they need to match the naming pattern but you really only need one so we use win-x64 and use it's version everywhere.

lewing commented on eng/Version.Details.xml:
```diff
@@ -414,5 +414,85 @@
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Sdk.win-x64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Sdk.win-arm64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Sdk.osx-x64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Sdk.osx-arm64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Cache.win-x64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Cache.win-arm64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Cache.linux-x64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Cache.linux-arm64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Cache.osx-x64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.Emscripten.3.1.34.Cache.osx-arm64" Version="9.0.0-preview.4.24175.6">
+      <Uri>https://github.com/dotnet/emsdk</Uri>
+      <Sha>2106e0604f52119b7d49577476ece7d51a23a40d</Sha>
+    </Dependency>
```
```suggestion
```
This isn't critical for the I believe but the transport package is enough to establish the dependency for darc and these package names will change with the emsdk bump and would require manual updating each time.

lewing reviewed (commented):
The versions.props suggestion needs to land but it looks like we are almost there

lewing committed:
Use a single version of emsdk and take node version from flow

lewing committed:
Merge branch 'main' into test-build-wasm

lewing committed:
Merge branch 'main' into test-build-wasm

radekdoulik committed:
Update eng/Version.Details.xml

Co-authored-by: Larry Ewing <lewing@microsoft.com>

radekdoulik committed:
Merge branch 'main' into test-build-wasm

akoeplinger commented on eng/Versions.props:
```diff
@@ -256,5 +256,9 @@
+    <runtimewinx64MicrosoftNETCoreRuntimeWasmNodeTransportPackageVersion>9.0.0-alpha.1.24175.1</runtimewinx64MicrosoftNETCoreRuntimeWasmNodeTransportPackageVersion>
+    <EmsdkPackageVersion>$(MicrosoftNETRuntimeEmscriptenVersion)</EmsdkPackageVersion>
+    <NodePackageVersion>$(runtimewinx64MicrosoftNETCoreRuntimeWasmNodeTransportPackageVersion)</NodePackageVersion>
+    <EmsdkVersion>3.1.34</EmsdkVersion>
```
please move these up next to MicrosoftNETWorkloadEmscriptenCurrentManifest90100TransportVersion

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -157,25 +173,133 @@
-      <InstallCmd Condition="'$(HostOS)' == 'windows'">powershell -NonInteractive -command &quot;&amp; $(InstallCmd); Exit $LastExitCode &quot;</InstallCmd>
-      <ActivateCmd Condition="'$(HostOS)' == 'windows'">powershell -NonInteractive -command &quot;&amp; $(ActivateCmd); Exit $LastExitCode &quot;</ActivateCmd>
-      <PythonCmd Condition="'$(HostOS)' == 'windows'and '$(TargetsBrowser)' == 'true'">setlocal EnableDelayedExpansion &amp;&amp; call &quot;$([MSBuild]::NormalizePath('$(EMSDK_PATH)', 'emsdk_env.bat'))&quot; &amp;&amp; !EMSDK_PYTHON!</PythonCmd>
+      <_EmsdkPaths Condition="'$(HostOS)' != 'windows'">
```
please move the scripts into separate files, they shouldn't be inside mono.proj

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -157,25 +173,133 @@
+                    Include="$(NuGetPackageRoot)\$([System.String]::Copy(%(PackageReference.Identity)).ToLowerInvariant())\%(PackageReference.Version)\tools\**" />
+      <NodeFiles    Condition="'%(PackageReference.Identity)' == 'runtime.$(_portableHostOS)-$(BuildArchitecture).Microsoft.NETCore.Runtime.Wasm.Node.Transport'"
+                    Include="$(NuGetPackageRoot)\$([System.String]::Copy(%(PackageReference.Identity)).ToLowerInvariant())\%(PackageReference.Version)\tools\$(_portableHostOS)-$(BuildArchitecture)\**" />
+      <PythonFiles  Condition="'$(HostOS)' == 'windows' and '%(PackageReference.Identity)' == 'Microsoft.NET.Runtime.Emscripten.3.1.34.Python.win-$(BuildArchitecture)'"
```
this shouldn't hardcode 3.1.34 here
